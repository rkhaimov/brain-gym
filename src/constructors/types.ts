import { Either } from '../either';

export type TypeConstructor<TLeft, TRight> = (
  value: unknown
) => Either<TLeft, TRight>;
