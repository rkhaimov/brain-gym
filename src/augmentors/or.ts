import { TypeNode } from '../core';
import { validate } from '../validate';
import { refine } from './refine';
import { unknown } from '../types/unknown';

export const or = <TLeft, TRight>(
  left: TypeNode<TLeft>,
  right: TypeNode<TRight>
): TypeNode<TLeft | TRight> =>
  refine(unknown(), {
    validate: (value) => {
      const lerrors = validate(left, value as TLeft);

      if (lerrors.length === 0) {
        return [];
      }

      const rerrors = validate(right, value as TRight);

      if (rerrors.length === 0) {
        return [];
      }

      return [...lerrors, ...rerrors];
    },
    defaults: left.defaults,
  });
