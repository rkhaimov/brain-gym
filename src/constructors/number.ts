import { TypeConstructor } from './types';
import { Either } from '../either';

export const number = (): TypeConstructor<string, number> => (value) => {
  if (typeof value === 'number') {
    return Either.right(value);
  }

  return Either.left('is not a number');
};

