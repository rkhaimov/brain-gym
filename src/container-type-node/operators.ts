// TODO: Should generalize to all structures (tuples, trees, graphs)
import { ToType, TypeNode, TypeNodeOperator } from '../core';
import { ContainerTypeNode } from './createContainerTypeNode';
import { andValidate, defaultsTo } from '../operators';
import { fromErrorMessage } from '../validate';

export const nonEmptyList =
  <TChildren extends TypeNode>(): TypeNodeOperator<
    ContainerTypeNode<ToType<TChildren>[], TChildren>
  > =>
  (tn) =>
    tn.wrap(
      defaultsTo(() => [tn.children().defaults() as ToType<TChildren>]),
      andValidate((value) => {
        if (value.length === 0) {
          return [fromErrorMessage('list is empty')];
        }

        return [];
      })
    );
