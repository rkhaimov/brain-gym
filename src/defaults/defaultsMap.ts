import { ToType, TypeNode, TypeNodeOperator } from '../core';

export const defaultsMap =
  <TTypeNode extends TypeNode>(
    transform: (defaults: ToType<TTypeNode>, tn: TTypeNode) => ToType<TTypeNode>
  ): TypeNodeOperator<TTypeNode> =>
  (tn) =>
    tn.__clone({
      defaults: () => transform(tn.defaults() as ToType<TTypeNode>, tn),
    });
