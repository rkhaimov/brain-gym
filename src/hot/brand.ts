import { TypeNode } from '../core';

export const brand = <TType, TBrand extends string>(
  tn: TypeNode<TType>,
  brand: TBrand
): TypeNode<TType & { __brand: TBrand }> =>
  tn as TypeNode<TType & { __brand: TBrand }>;
