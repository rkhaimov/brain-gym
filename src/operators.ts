import { ToType, TypeNode, TypeNodeOperator } from './core';

export const defaultsTo =
  <TTypeNode extends TypeNode>(
    defaults: () => ToType<TTypeNode>
  ): TypeNodeOperator<TTypeNode> =>
  (tn) =>
    tn.clone({ validate: tn.validate, defaults });

export const andValidate =
  <TTypeNode extends TypeNode>(
    validate: TTypeNode['validate']
  ): TypeNodeOperator<TTypeNode> =>
  (tn) =>
    tn.clone({
      validate: (value) => [...tn.validate(value), ...validate(value)],
      defaults: tn.defaults,
    });
