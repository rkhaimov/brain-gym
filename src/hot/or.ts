import { TypeNode } from '../core';
import { validate } from '../validate';
import { unknown } from '../types/unknown';
import { refineMap } from '../operators/refineMap';
import { defaultsFrom } from '../operators/defaultsFrom';

export const or = <TLeft, TRight>(
  left: TypeNode<TLeft>,
  right: TypeNode<TRight>
) =>
  unknown().pipe(
    refineMap((value: TLeft | TRight) => {
      const lerrors = validate(left, value as TLeft);

      if (lerrors.length === 0) {
        return [];
      }

      const rerrors = validate(right, value as TRight);

      if (rerrors.length === 0) {
        return [];
      }

      return [...lerrors, ...rerrors];
    }),
    defaultsFrom(left)
  );
