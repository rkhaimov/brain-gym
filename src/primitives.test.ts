import { defaultsMap } from './defaultsMap';
import { or } from './or';
import { translateTo } from './translateTo';
import { brand } from './type-node/HOT/brand';
import { biggerOrEqThan } from './type-node/operators/biggerOrEqThan';
import { dividableBy } from './type-node/operators/dividableBy';
import { lesserOrEqThan } from './type-node/operators/lesserOrEqThan';
import { nil } from './type-node/types/nil';
import { number } from './type-node/types/number';
import { defaults, narrow, validate } from './validate';

describe('Primitive types', () => {
  describe('Validates and generates defaults', () => {
    it('validates with static checking', () => {
      expectNoErrors(validate(number(), 42));

      // @ts-expect-error // Static check
      expect(validate(number(), '42')).toMatchSnapshot();
    });

    it('narrows in assertion like manner', () => {
      const numeric: unknown = 42;

      expect(narrow(number(), numeric)).toEqual(numeric);

      expect(() => narrow(number(), '42')).toThrowErrorMatchingSnapshot();
    });

    it('generates defaults which must hold the following property', () => {
      const tn = number();

      expect(validate(tn, tn.defaults())).toEqual([]);
    });

    it('allows to override defaults', () => {
      const numeric = 10;
      const tn = number().wrap(defaultsMap(() => numeric));

      expect(tn.defaults()).toEqual(numeric);
    });

    it('allows to override error message', () => {
      const tn = number().wrap(
        biggerOrEqThan(10),
        dividableBy(2),
        translateTo({
          biggerOrEqThan: (n) => `Should be greater than ${n}`,
          dividableBy: (n) => `Should be dividable by ${n}`,
        })
      );

      expect(validate(tn, 2)).toMatchSnapshot();

      expect(validate(tn, 13)).toMatchSnapshot();

      const rtn = tn.wrap(
        translateTo({
          dividableBy: (n) => `Please, provide value dividable by ${n}`,
        })
      );

      expect(validate(rtn, 13)).toMatchSnapshot();
      expect(validate(tn.wrap(dividableBy(5)), 12)).toMatchSnapshot();
    });
  });

  describe('Constraints composition', () => {
    it('allows to add additional constraints', () => {
      const tn = number().wrap(biggerOrEqThan(50));

      expect(validate(tn, 20)).toMatchSnapshot();

      // @ts-expect-error // Static check
      expect(validate(tn, '60')).toMatchSnapshot();

      // @ts-expect-error // Static check
      expect(validate(tn, '42')).toMatchSnapshot();

      expectNoErrors(validate(tn, 100));
    });

    it('generates defaults for composite types', () => {
      const tn = number().wrap(biggerOrEqThan(50));

      expect(validate(tn, tn.defaults())).toEqual([]);
    });

    it('throws non matching defaults error', () => {
      const tn = number().wrap(biggerOrEqThan(5)).wrap(lesserOrEqThan(3));

      expect(() => defaults(tn)).toThrowErrorMatchingSnapshot();
    });

    it('allows to provide manual defaults to bypass error', () => {
      const tn = number().wrap(
        dividableBy(2),
        dividableBy(3),
        defaultsMap(() => 2 * 3)
      );

      expect(validate(tn, tn.defaults())).toEqual([]);
    });
  });

  describe('Type widening', () => {
    it('primitive types can be nullable', () => {
      const tn = or(number(), nil());

      // @ts-expect-error // Static check
      expect(validate(tn, '42')).toMatchSnapshot();

      expectNoErrors(validate(tn, undefined));

      expectNoErrors(validate(tn, 42));
    });

    it('defaults are generated on first value by default and can be customized on demand', () => {
      const tn0 = number();
      const tn1 = nil();
      const tn = or(tn0, tn1);

      expect(tn.defaults()).toEqual(tn0.defaults());

      const customized = tn.wrap(defaultsMap(() => tn1.defaults()));

      expect(customized.defaults()).toEqual(tn1.defaults());
    });
  });

  describe('Type narrowing', () => {
    it('Primitive types can be branded', () => {
      const tn = brand(number(), 'id');

      expectNoErrors(validate(tn, tn.defaults()));
      expectNoErrors(validate(tn, narrow(tn, 42)));
    });
  });
});

function expectNoErrors(errors: unknown[]) {
  expect(errors.length).toBe(0);
}
