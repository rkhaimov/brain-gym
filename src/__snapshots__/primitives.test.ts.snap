// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Primitive types Constraints composition allows to add additional constraints 1`] = `
Array [
  Object {
    "kind": "biggerOrEqThan",
    "message": "biggerOrEqThan: 50",
    "params": 50,
    "paths": Array [],
  },
]
`;

exports[`Primitive types Constraints composition allows to add additional constraints 2`] = `
Array [
  Object {
    "kind": "number",
    "message": "number: no params",
    "params": undefined,
    "paths": Array [],
  },
]
`;

exports[`Primitive types Constraints composition allows to add additional constraints 3`] = `
Array [
  Object {
    "kind": "number",
    "message": "number: no params",
    "params": undefined,
    "paths": Array [],
  },
]
`;

exports[`Primitive types Constraints composition throws non matching defaults error 1`] = `"Empty path: biggerOrEqThan: 5"`;

exports[`Primitive types Type widening primitive types can be nullable 1`] = `
Array [
  Object {
    "kind": "number",
    "message": "number: no params",
    "params": undefined,
    "paths": Array [],
  },
  Object {
    "kind": "nil",
    "message": "nil: no params",
    "params": undefined,
    "paths": Array [],
  },
]
`;

exports[`Primitive types Validates and generates defaults allows to override error message 1`] = `
Array [
  Object {
    "kind": "biggerOrEqThan",
    "message": "Should be greater than 10",
    "params": 10,
    "paths": Array [],
  },
]
`;

exports[`Primitive types Validates and generates defaults allows to override error message 2`] = `
Array [
  Object {
    "kind": "dividableBy",
    "message": "Should be dividable by 2",
    "params": 2,
    "paths": Array [],
  },
]
`;

exports[`Primitive types Validates and generates defaults allows to override error message 3`] = `
Array [
  Object {
    "kind": "dividableBy",
    "message": "Please, provide value dividable by 2",
    "params": 2,
    "paths": Array [],
  },
]
`;

exports[`Primitive types Validates and generates defaults allows to override error message 4`] = `
Array [
  Object {
    "kind": "dividableBy",
    "message": "Should be dividable by 5",
    "params": 5,
    "paths": Array [],
  },
]
`;

exports[`Primitive types Validates and generates defaults narrows in assertion like manner 1`] = `"Empty path: number: no params"`;

exports[`Primitive types Validates and generates defaults validates with static checking 1`] = `
Array [
  Object {
    "kind": "number",
    "message": "number: no params",
    "params": undefined,
    "paths": Array [],
  },
]
`;
